### PR TITLE
Adds SharedPtr Get interface for Pos & Vel solvers

### PR DESCRIPTION
**Pull request by rethink-imcmahon**
This Adds a shared_ptr get interface for the position and velocity solvers inside the SNS_IK class.

Also, it fills out the testing framework a bit by iterating over different types of velocity solvers.

### CODE REVIEW

File: sns_ik_examples/src/ik_tests.cpp [hunk 1]
```diff
@@ -209,6 +210,37 @@ void test(ros::NodeHandle& nh, double num_samples_pos, double num_samples_vel, s
...
   // SNS Position Tests
+  struct velocitySolverData {
+    sns_ik::VelocitySolveType type;
+    std::string               name;
+    double                   score;
+    double           scaling_score;
+    double                avg_time;
+  };
+  std::vector<velocitySolverData> vel_solver_data;
+  velocitySolverData sns = {sns_ik::SNS,"SNS",0.0,0.0};
```
Comment (rethink-forrest) [inline sns_ik_examples/src/ik_tests.cpp]: The avg_time is not being set, though it probably defaults to zero.

File: sns_ik_examples/src/ik_tests.cpp [hunk 2]
```diff
@@ -209,6 +210,37 @@ void test(ros::NodeHandle& nh, double num_samples_pos, double num_samples_vel, s
...
+  velocitySolverData sns_optimalsm = {sns_ik::SNS_OptimalScaleMargin,"SNS Optimal Scale Margin",0.0,0.0};
+  vel_solver_data.push_back(sns_optimalsm);
+  velocitySolverData sns_optimal = {sns_ik::SNS_Optimal,"SNS Optimal",0.0,0.0};
+  vel_solver_data.push_back(sns_optimal);
+  velocitySolverData sns_fast = {sns_ik::SNS_Fast,"SNS Fast",0.0,0.0};
+  vel_solver_data.push_back(sns_fast);
+  velocitySolverData sns_fastoptimal = {sns_ik::SNS_FastOptimal,"SNS Fast Optimal",0.0,0.0};
+  vel_solver_data.push_back(sns_fastoptimal);
+  double linearTolerance = 0.001;
+  double angularTolerance = 0.001;
```
Comment (rethink-forrest) [inline sns_ik_examples/src/ik_tests.cpp]: The tolerances should be kept at 1e-5 to match the tolerances used by kdl and trac-ik.

File: sns_ik_examples/src/ik_tests.cpp [hunk 3]
```diff
@@ -209,6 +210,37 @@ void test(ros::NodeHandle& nh, double num_samples_pos, double num_samples_vel, s
...
+  double linearMaxStepSize = 0.001;
+  double angularMaxStepSize = 0.001;
+  double maxIterations = 5;
+  double dt=0.01;
```
Comment (rethink-forrest) [inline sns_ik_examples/src/ik_tests.cpp]: Make a note that these aren't vaild values yet, or set to something like:

double posIK_linearMaxStepSize = 0.05;
double posIK_angularMaxStepSize = 0.05;
double posIK_maxIterations = 150;
double posIK_dt=0.2;

File: sns_ik_examples/src/ik_tests.cpp [hunk 4]
```diff
@@ -209,6 +210,37 @@ void test(ros::NodeHandle& nh, double num_samples_pos, double num_samples_vel, s
...
+  for(auto& vst: vel_solver_data){
+  snsik_solver.setVelocitySolveType(vst.type);
```
Comment (rethink-forrest) [inline sns_ik_examples/src/ik_tests.cpp]: Indent for loop.

File: sns_ik_examples/src/ik_tests.cpp [hunk 5]
```diff
@@ -225,12 +257,18 @@ void test(ros::NodeHandle& nh, double num_samples_pos, double num_samples_vel, s
     if (int((double)i/num_samples_pos*100)%10 == 0)
       ROS_INFO_STREAM_THROTTLE(1,int((i)/num_samples_pos*100)<<"\% done");
   }
-  ROS_INFO_STREAM("SNS-IK found "<<success<<" solutions ("<<100.0*success/num_samples_pos<<"\%) with an average of "<<total_time/num_samples_pos<<" secs per sample");
-
+  vst.score=100.0*success/num_samples_pos;
+  vst.avg_time=total_time/num_samples_pos;
+  ROS_INFO_STREAM(vst.name<<" found "<<success<<" solutions ("<<vst.score<<"\%) with an average of "<<vst.avg_time<<" secs per sample");
```
Comment (rethink-forrest) [inline sns_ik_examples/src/ik_tests.cpp]: Include the success with scaling results, or I could add that after this merge. Also, watch your line length....

File: sns_ik_lib/include/sns_ik/sns_position_ik.hpp [hunk 6]
```diff
@@ -47,14 +47,25 @@ class SNSPositionIK {
...
+    void setStepSize(double linearMaxStepSize, double angularMaxStepSize);
+    void setSolveTimeParameters(double dt, double maxIterations);
 
   private:
     KDL::Chain m_chain;
     std::shared_ptr<SNSVelocityIK> m_ikVelSolver;
     KDL::ChainFkSolverPos_recursive m_positionFK;
     KDL::ChainJntToJacSolver m_jacobianSolver;
+    double m_linearTolerance;
+    double m_angularTolerance;
```
Comment (rethink-forrest) [inline sns_ik_lib/include/sns_ik/sns_position_ik.hpp]: The tolerances don't need to be member variables.

File: sns_ik_lib/src/sns_position_ik.cpp [hunk 7]
```diff
@@ -35,28 +41,31 @@ SNSPositionIK::~SNSPositionIK()
...
 {
-  m_chain = chain;
-  m_positionFK = KDL::ChainFkSolverPos_recursive(chain);
-  m_jacobianSolver = KDL::ChainJntToJacSolver(chain);
+  m_linearMaxStepSize = linearMaxStepSize;
+  m_angularMaxStepSize = angularMaxStepSize;
 }
-*/
+
+void SNSPositionIK::setSolveTimeParameters(double dt, double maxIterations)
```
Comment (rethink-forrest) [inline sns_ik_lib/src/sns_position_ik.cpp]: If anything, I would group dt with the step sizes and not with maxIterations.

File: sns_ik_examples/src/ik_tests.cpp (hunk 1 above)
Comment (rethink-imcmahon) [inline sns_ik_examples/src/ik_tests.cpp]: Oh, interesting. That's because I added `scaling_score` field afterward. I'll modify the code.

**Comment by rethink-imcmahon** [comment]
Updated based on your comments, Forrest.

**Comment by rethink-forrest** [comment]
Did you build after after your latest commit? Looks like you might have broken the build. 

Specifically:
* pos_solver->setTolerance used in ik_tests.cpp no long exists
* SNSPositionIK::getChain() fails because you try to directly copy a KDL::Chain into std::shared_ptr<KDL::Chain>&

**Comment by rethink-imcmahon** [comment]
Right you are. I thought I did test it, but apparently not.

**Pull request merged**